Migrating PGO CLI kuttl tests to chainsaw

philrhurst opened the pull request:
These are the first round of kuttl tests migrated to chainsaw. The structure is a single test for each kubectl-pgo command

- backup
- create
- delete
- show
- version

philrhurst committed:
Kyverno policies

We can use Kyverno policies to facilitate testing with different image repos.

philrhurst committed:
initial set of templates

These are helpful templates to use when testing the PGO CLI

philrhurst committed:
default config and values YAML files

philrhurst committed:
Chainsaw test for PGO CLI backup

philrhurst committed:
Chainsaw test for PGO CLI backup

philrhurst committed:
Chainsaw tests for PGO CLI create

philrhurst committed:
Chainsaw tests for PGO CLI delete

philrhurst committed:
Chainsaw test for PGO CLI show

philrhurst committed:
Chainsaw tests for PGO CLI version

philrhurst committed:
Add Chainsaw README

Instructions for how to run the tests

philrhurst committed:
remove skipDelete

The chainsaw-show Test should remove resources after it completes.

philrhurst committed:
Chainsaw test for PGO CLI start-stop

philrhurst committed:
Chainsaw test for PGO CLI restore

philrhurst committed:
Update chainsaw tests with helpful sleeps

philrhurst committed:
Update template with better check for backup complete

tjmoore4 reviewed (commented):
Looks good overall. A few questions.

benjaminjb commented on testing/chainsaw/e2e/backup/chainsaw-test.yaml:
```diff
@@ -0,0 +1,705 @@
+    - name: Sleep for 10s
+      try:
+      - sleep:
+          duration: 10s
```
Can we set timeouts on different steps to give longer timeouts for the confirm steps or whatever that is taking so long.

benjaminjb commented on testing/chainsaw/e2e/create/chainsaw-test.yaml:
```diff
@@ -0,0 +1,146 @@
+      use:
+        template: '../templates/confirm-created.yaml'
+
+    - name: Sleep for 30s
```
If we need these sleep steps, I think it would be fine to name them all `Sleep` or something without the time -- it's easy enough to look at the duration to see how long and it avoids drift when we change those durations.

More interesting to me would be why we chose those durations.

philrhurst replied:
Most of the Sleep 30s started at Sleep 10. I bumped them up because running all the test simultaneously was causing issues. I left the descriptive "Sleep for 30s" as I was debugging the tests to know how much time I should wait before the next step.

benjaminjb commented on testing/chainsaw/e2e/start-stop/chainsaw-test.yaml:
```diff
@@ -0,0 +1,255 @@
+
+    - name: 'Confirm cluster is created'
+      use:
+        template: '../templates/confirm-created.yaml'
```
So the idea is that we can use the same template here because the command did nothing?

benjaminjb commented on testing/chainsaw/e2e/start-stop/chainsaw-test.yaml:
```diff
@@ -0,0 +1,255 @@
+        - sleep:
+            duration: 30s
+
+    - name: run 'stop cluster' with confirm 'y' and force-conflicts
```
if the `start` test runs the stop command, I think we can collapse this into one test, yeah?

benjaminjb commented on testing/chainsaw/policies/v1.15.1/00-rbac.yaml:
```diff
@@ -0,0 +1,32 @@
+  verbs:
+  - create
+  - update
+  - delete
```
Why delete?

philrhurst replied:
I was having some issues getting it to work. This Github comment suggested delete.

- https://github.com/kyverno/kyverno/issues/11467#issuecomment-2480926228

philrhurst commented on testing/chainsaw/e2e/backup/chainsaw-test.yaml:
```diff
@@ -0,0 +1,705 @@
+    - name: Verify COMMAND_OPTS matches "--stanza=db --repo=1"
+      bindings:
+        - name: selector
+          value: (replace_all(
```
yes - it was a bit of over-engineering. The backupSelector is constant for the entire test; so we can just define it fully.

philrhurst replied:
I've refactored to simplify the select injection. No more nesting!

philrhurst commented on testing/chainsaw/e2e/backup/chainsaw-test.yaml:
```diff
@@ -0,0 +1,705 @@
+apiVersion: chainsaw.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: check-backup-command-no-flags
```
This is the conversion of `08--backup-with-just-trigger.yaml` which is triggering a backup from the PGO CLI with no options and checking the annotations

philrhurst commented on testing/chainsaw/e2e/restore/chainsaw-test.yaml:
```diff
@@ -0,0 +1,713 @@
+apiVersion: chainsaw.kyverno.io/v1alpha1
+kind: Test
+metadata:
+  name: check-restore-confirm-yes
```
Yes it is, but each "test" would then be a separate "step". It becomes an issue of clarity and preference. I.e. do we have a single Restore Test that does multiple things. Or multiple Restore Tests, each doing one single thing.

One thing I like about Chainsaw is that each test is namespace-scoped. So it's easy to separate out tests into discrete units.

philrhurst commented on testing/chainsaw/e2e/restore/chainsaw-test.yaml:
```diff
@@ -0,0 +1,713 @@
+            (contains($stderr, 'Apply failed')): true
+            (contains($stderr, '2 conflicts')): true
+
+    - name: run 'restore' with confirm 'yes' and --force-conflicts
```
yes, I will look into this. I think I cribbed more of the original kuttl test on this particular one than the others.

philrhurst commented on testing/chainsaw/e2e/values.yaml:
```diff
@@ -0,0 +1,5 @@
+  postgres: '17'
+
+images:
+  psql: 'registry.developers.crunchydata.com/crunchydata/crunchy-postgres:ubi9-17.5-2520'
```
It probably should be - I was basing this on an example we were using elsewhere.

philrhurst commented on testing/chainsaw/e2e/version/chainsaw-test.yaml:
```diff
@@ -0,0 +1,50 @@
+  name: version
+spec:
+  steps:
+  - name: step-00
```
sure :)

philrhurst committed:
added newlines at the end

philrhurst committed:
renamed steps in version test

philrhurst committed:
refactored backupSelector

The way we defined the selector for backups was quite terse. This refactor simplifies the logic.

philrhurst committed:
Chainsaw test for PGO CLI support export

tjmoore4 reviewed (approved):
Updates look good to me. Not sure if @benjaminjb is still tracking any blockers, but I think this is good to merge.

benjaminjb commented on testing/chainsaw/e2e/support/chainsaw-test.yaml:
```diff
@@ -0,0 +1,932 @@
+    - name: cluster
+      value: kuttl-support-cluster
+
+    - name: postgresVersion
+      value: 16
```
Don't we have a central source for pg version?

philrhurst replied:
Yes we do; but the bash for this particular test has already hard-coded 16 in it. I didn't want to refactor to interrupt migrating the tests.

benjaminjb commented on testing/chainsaw/e2e/support/chainsaw-test.yaml:
```diff
@@ -0,0 +1,932 @@
+            metadata:
+              name: kuttl-support-monitoring-cluster
+            spec:
+              postgresVersion: 16
```
and if we have a central location for the pg version (or even a local version), better to use it

benjaminjb commented on testing/chainsaw/e2e/support/chainsaw-test.yaml:
```diff
@@ -0,0 +1,932 @@
+                      requests:
+                        storage: 1Gi
+                  replicas: 1
+              postgresVersion: 16
```
ditto

benjaminjb reviewed (approved):
Really no blockers, but re-reading these tests -- many of which I probably had a hand in -- makes me want to revise these tests. Please make a ticket so we don't forget to!

(What I really want is to shift tests left, e.g., run as much as we can as unit tests) and also combine a few of these tests if we can to run them faster. I know I have created a few tickets re: separating out the cluster work from some CLI work, so it won't just be test changes I'm envisioning in the future.)
